### PR TITLE
fix: use @ipld/dag-pb instead of ipld-dag-pb

### PR DESCRIPTION
**Pull request by vmx**
Use the rewrite of the dag-pb codec.

### CODE REVIEW

File: packages/ipfs-unixfs-importer/package.json [hunk 1]
```diff
@@ -40,6 +40,7 @@
     "detect-node": "^2.0.4",
     "ipfs-unixfs-exporter": "^3.0.7",
     "ipld": "^0.28.0",
+    "ipld-dag-pb": "^0.20.0",
```
Comment (rvagg) [inline packages/ipfs-unixfs-importer/package.json:43]: this isn't needed in here still is it?

File: packages/ipfs-unixfs-importer/src/dag-builder/file/buffer-importer.js [hunk 2]
```diff
@@ -27,7 +28,7 @@ async function * bufferImporter (file, source, block, options) {
           mode: file.mode
         })
 
-        buffer = new DAGNode(unixfs.marshal()).serialize()
+        buffer = encode(prepare({ Data: unixfs.marshal() }))
```
Comment (rvagg) [inline packages/ipfs-unixfs-importer/src/dag-builder/file/buffer-importer.js:31]: `prepare()` has some super-powers - if you just give it a `Uint8Array` it'll do `{Data: data}` for you, so you could trim this line down even further

File: packages/ipfs-unixfs-importer/src/dag-builder/file/index.js [hunk 3]
```diff
@@ -66,7 +66,7 @@ const reduce = (file, block, options) => {
         })
 
         const multihash = mh.decode(leaf.cid.multihash)
-        buffer = new DAGNode(leaf.unixfs.marshal()).serialize()
+        buffer = encode(prepare({ Data: leaf.unixfs.marshal() }))
```
Comment (rvagg) [inline packages/ipfs-unixfs-importer/src/dag-builder/file/index.js:69]: ditto on shortcut

**Comment by rvagg** [comment]
Surprisingly easy, maybe deceptively so? I hope not!
The only catch might be `Tsize` as discussed in https://github.com/ipld/js-dag-pb/issues/5, either we don't have test cases that assert expected CIDs for cases where `Tsize` might be defaulted to `0` in the old stack but missing in the new, or `@ipld/dag-pb` is doing the same thing by omitting it. I don't _think_ this is the case though, the old JS library seems to be defaulting to `0` where you omit it. So we should resolve this. Either just make `prepare()` more of a hack by adding it in. Or, ideally, fix it at the spec level so we have consistent behaviour between Go and JS (and other) implementations.

**Comment by achingbrain** [comment]
This is great, thanks for submitting it.  Could you please open a PR to js-ipfs that uses this branch?  Before merging it would be great to see the effect on bundle size, performance, and compatibility.  If it's 'not much' then it shows we can roll this upgrade out piecemeal instead of in a big bang.

File: packages/ipfs-unixfs-importer/package.json (hunk 1 above)
Comment (vmx) [reply packages/ipfs-unixfs-importer/package.json:43]: It is as a dev dependency as the tests use js-ipld, which expects data to come as and `ipld-dag-pb` `DAGNode`.

File: packages/ipfs-unixfs-importer/src/dag-builder/file/buffer-importer.js (hunk 2 above)
Comment (vmx) [reply packages/ipfs-unixfs-importer/src/dag-builder/file/buffer-importer.js:31]: I thought about it, but when I then read the code I actually couldn't make sense of what it is doing anymore. I hope that this indicates clearer that we are creating a DAG-PB node here.

**Comment by rvagg** [comment]
I came back here to comment on the js-multiformats `CID` vs js-cid `CID` problem arising out of `@ipld/dag-pb` decoding, but it looks like this PR has expanded to a full `CID` replacement effort! How's that going @vmx?

**Comment by lidel** [comment]
Feedback from triage: TS needs to land before this.

**Comment by rvagg** [comment]
Rebased the original branch to `master` and have now pulled in the WIP [`multiformats`](https://github.com/multiformats/js-multiformats/pull/75), [`@ipld/dag-pb`](https://github.com/ipld/js-dag-pb/pull/6) and [`@ipld/dag-cbor`](https://github.com/ipld/js-dag-cbor/pull/18), removed the interim `PbNode` and `PbLink` and a bunch of other tweaks now we have proper types coming in.

CI is passing except for FF and Chrome which are all failing with `ReferenceError: chai is not defined`, is this a known issue with the current aegir? Are we pulling it in the right way?

**Comment by rvagg** [comment]
HAH! Fixed CI and I have to relay how I fixed it because it's 🤯

I'm trying to fix the `--browser` tests that all fail with [bundle load errors](https://travis-ci.com/github/ipfs/js-ipfs-unixfs/jobs/500709784):

```
ipfs-unixfs: ReferenceError: chai is not defined
ipfs-unixfs:     at file:///home/travis/build/ipfs/js-ipfs-unixfs/node_modules/dirty-chai/lib/dirty-chai.js:15:7
ipfs-unixfs:     at file:///home/travis/build/ipfs/js-ipfs-unixfs/node_modules/dirty-chai/lib/dirty-chai.js:15:16
ipfs-unixfs:     at http://127.0.0.1:3000/bundle-out.js:3:47
ipfs-unixfs:     at file:///home/travis/build/ipfs/js-ipfs-unixfs/node_modules/aegir/utils/chai.js:7:10
ipfs-unixfs:     at http://127.0.0.1:3000/bundle-out.js:3:47
ipfs-unixfs:     at file:///home/travis/build/ipfs/js-ipfs-unixfs/packages/ipfs-unixfs/test/unixfs-format.spec.js:4:20
ipfs-unixfs:     at http://127.0.0.1:3000/bundle-out.js:3:47
ipfs-unixfs:     at file:///in.js:8:1
ipfs-unixfs:     at http://127.0.0.1:3000/bundle-out.js:33001:3
```

dirty-chai uses old school environment detection to figure out how to inject, so its CJS detection looks like this:

```js
if (typeof require === 'function' && typeof exports === 'object' && typeof module === 'object') { ... }
```

**esbuild** translates this as:

```js
if (typeof exports2 === "object" && typeof module2 === "object") { ... }
```

and passes in `exports2` and `module2`, so all good.

At least this is what it does for me locally when trying to match the Travis environment -- Linux x64 Node.js v14.16.1 and npm@latest (updated as part of the Travis script here).

But looking at what's going on in the bundle by modifying the Travis script with:

```yml
      script:
        - sed  's/  premove/  \/\/premove/' -i node_modules/playwright-test/src/runner.js
        - npm run test -- -- -- -t browser || true
        - grep dirty-chai -C 5 /tmp/*/bundle-out.js
```

We can see that the CJS detection is this (see [here](https://travis-ci.com/github/ipfs/js-ipfs-unixfs/jobs/500709783)):

```js
if (typeof require === "function" && typeof exports2 === "object" && typeof module2 === "object") { ... }
```

**BUT** there's one tiny difference between my local env and Travis -- Travis is using Xenial (16.04) and I'm on Focal (20.04). So in the latest commit I've included `dist: focal` and ✅ , we're working now!

I just can't explain _why_. esbuild should download [esbuild-linux-64](https://www.npmjs.com/package/esbuild-linux-64) for its binary and this should be the same between both dists. There must be something slightly different in the way it executes on 16.04 vs 20.04 in this environment. The change introduced here that's likely triggering this is that some of the dependency tree can now be consumed as ESM if the bundler so chooses. That must be invoking some code-path that's active on an older libc system than a newer one .. yay Go? Or maybe there's a more straightfoward explanation.

**Comment by BigLep** [comment]
Assigned to @achingbrain given he has the context to make meaningful progress in a tribute week.  It can get picked up when he is on Tribute rotation.

**Review by Gozala — APPROVED** [review]
Other than question already addressed at @rvagg (about nameless links) it looks good to me.

**Comment by rvagg** [comment]
@achingbrain @Gozala I've pushed a commit to this branch that uses the new `createNode()` and `createLink()` factories from dag-pb as per https://github.com/ipld/js-dag-pb/pull/20. After implementing it here it seems like a win to me - less magic property name futzing and fewer lines of code. It almost retires the need for `prepare()` in fact. If you think it's a negative then let me know, it can be backed out.

8e9a01c178fa1ce67f40c3becef0accb3cce6d5b

**Comment by lidel** [comment]
Triage note: waiting for green CI for js-ipfs

**Review by achingbrain — APPROVED** [review]

**Comment by olizilla** [comment]
GO TEAM 🎉

**Pull request merged**